src/coreclr/jit/helperexpansion.cpp

```diff
@@ -443,7 +443,282 @@ PhaseStatus Compiler::fgExpandThreadLocalAccess()
+    DISPTREE(call);
+    JITDUMP("\n");
+    bool hasLazyStaticCtor = call->IsExpTLSFieldAccessLazyCtor();
+
+    call->ClearExpTLSFieldAccess();
+    call->ClearExpTLSFieldAccessLazyCtor();
+
+    CORINFO_THREAD_STATIC_INFO_READYTORUN threadStaticInfo;
+    memset(&threadStaticInfo, 0, sizeof(CORINFO_THREAD_STATIC_INFO_READYTORUN));
+
+    info.compCompHnd->getThreadLocalStaticInfo_ReadyToRun(&threadStaticInfo, call->gtInitClsHnd);
+
+    JITDUMP("tlsRootObject= %p\n", dspPtr(threadStaticInfo.tlsRootObject.addr));
+    JITDUMP("tlsIndexObject= %p\n", dspPtr(threadStaticInfo.tlsIndexObject.addr));
+    JITDUMP("offsetOfThreadLocalStoragePointer= %u\n", dspOffset(threadStaticInfo.offsetOfThreadLocalStoragePointer));
+    JITDUMP("threadStaticBaseSlow= %p\n", dspPtr(threadStaticInfo.threadStaticBaseSlow.addr));
+    JITDUMP("classCtorContextSize= %u\n", threadStaticInfo.classCtorContextSize);
+    JITDUMP("lazyCtorRunHelper= %p\n", dspPtr(threadStaticInfo.lazyCtorRunHelper.addr));
+    JITDUMP("lazyCtorTargetSymbol= %p\n", dspPtr(threadStaticInfo.lazyCtorTargetSymbol.addr));
+
+    // Split block right before the call tree
+    BasicBlock* prevBb       = block;
+    GenTree**   callUse      = nullptr;
+    Statement*  newFirstStmt = nullptr;
+    DebugInfo   debugInfo    = stmt->GetDebugInfo();
+    block                    = fgSplitBlockBeforeTree(block, stmt, call, &newFirstStmt, &callUse);
+    *pBlock                  = block;
+    var_types callType       = call->TypeGet();
+    assert(prevBb != nullptr && block != nullptr);
+
+    BasicBlock* targetSymbCondBB = nullptr;
+    BasicBlock* lazyCtorBB       = nullptr;
+
+    unsigned finalLclNum         = lvaGrabTemp(true DEBUGARG("Final offset"));
+    lvaTable[finalLclNum].lvType = TYP_I_IMPL;
+    GenTree* finalLcl            = gtNewLclVarNode(finalLclNum);
+
+    // if (lazyCtorBlock)
+    // {
+    // targetSymbCondBB (BBJ_COND):                                     [weight: 1.0]
+    //      targetSymbolAddr = [targetSymbol + size]
+    //      if (*targetSymbolAddr == 0)
+    //          goto tlsRootNullCondBB;
+    //
+    // lazyCtorBlock (BBJ_ALWAYS):                                      [weight: 0]
+    //      tlsRoot = HelperCall(0, -1, targetSymbolAddr);
+    //      goto block;
+    //
+    // }
+    // tlsRootNullCondBB (BBJ_COND):                                    [weight: 1.0]
+    //      fastPathValue = [tlsRootAddress]
+    //      if (fastPathValue != nullptr)
+    //          goto fastPathBb;
+    //
+    // fallbackBb (BBJ_ALWAYS):                                         [weight: 0]
+    //      tlsRoot = HelperCall(tlsRootAddress);
+    //      goto block;
+    //
+    // fastPathBb(BBJ_ALWAYS):                                          [weight: 1.0]
+    //      tlsRoot = fastPathValue;
+    //
+    //
+    // block (...):                                                     [weight: 1.0]
+    //      use(tlsRoot);
+
+    if (hasLazyStaticCtor)
```

Comment:
> Would it be better to deal with the static constructor independently from the field access?

The expansion happens after all the optimizations, CSE, etc. is done. We tried earlier to do in earlier phase, but it was making it hard to hoist certain checks as mentioned in https://github.com/dotnet/runtime/pull/82973#issuecomment-1457688325. Ideally, we need to expand just subset of appropriate portion of tree in various phases to achieve idle code, but might be tricky to get all the cases. Even today, without these changes, we do not hoist the call to helper out of loop because of https://github.com/dotnet/runtime/blob/bd83e17052d3c09022bad1d91dca860ca6b27ab9/src/coreclr/jit/optimizer.cpp#L7575-L7580

```c#
    static void Test(int n)
    {
        for (int i = 0; i < n; i++)
        {
            x1 = 5; x2 = 5; x3 = 5;
            y1 = 5; y2 = 5; y3 = 5;
        }
    }
```

```asm
G_M000_IG01:                ;; offset=0000H
       push     rsi
       push     rbx
       sub      rsp, 40
       mov      ebx, ecx

G_M000_IG02:                ;; offset=0008H
       lea      rax, [(reloc 0x42aa78)]
       cmp      qword ptr [rax-08H], 0
       jne      SHORT G_M000_IG06

G_M000_IG03:                ;; offset=0016H
       xor      esi, esi
       test     ebx, ebx
       jle      SHORT G_M000_IG05

G_M000_IG04:                ;; offset=001CH
       call     CORINFO_HELP_READYTORUN_THREADSTATIC_BASE
       mov      dword ptr [rax+08H], 5
       mov      dword ptr [rax+0CH], 5
       mov      dword ptr [rax+10H], 5
       mov      dword ptr [rax+14H], 5
       mov      dword ptr [rax+18H], 5
       mov      dword ptr [rax+1CH], 5
       inc      esi
       cmp      esi, ebx
       jl       SHORT G_M000_IG04

G_M000_IG05:                ;; offset=0051H
       add      rsp, 40
       pop      rbx
       pop      rsi
       ret

G_M000_IG06:                ;; offset=0058H
       call     CORINFO_HELP_READYTORUN_NONGCSTATIC_BASE
       jmp      SHORT G_M000_IG03
```

I propose to take these changes for .NET 8 because of time constraints and revisit missing gaps in .NET 9.

Reply:
@dotnet/jit-contrib Do you believe that this duplication is worth it for the time being?

> .NET 8 because of time constraints and revisit missing gaps in .NET 9.

I think it may be too late to take this change for .NET 8. We are actively trying to shut things down for .NET 8.

Reply:
> I think it may be too late to take this change for .NET 8.

In that case, I can start working on resolving the duplication problem that you pointed and get this in for .NET 9.

Reply:
I tried exploring on the idea of expanding lazy static ctor separately and it could be beneficial in cases where we have both a static init and a TLS, defined and used on same type (see example below). But it still won't eliminate the extra ctor check we introduce by expanding static initializer. If we have the static ctor trigger early, the VN will not 
give it same number because under the hood, they call different helper. We could theortically avoid emitting the 2nd invocation of `ctor` check (`0x40000000004203c8` related check below) it during expansion, if we have seen similar ctor invocation earlier, but for that we will have to make sure the previous ctor invocation dominates the block having the 2nd ctor invocation.

```c#
static string i = 5.ToString();
[ThreadStatic]
static string j = 5.ToString();
void Foo()
{
    Consume(i);
    Consume(j);
}
```

```asm
G_M27646_IG02:  ;; offset=0x0005
       lea      rcx, [(reloc 0x40000000004203c8)]
       cmp      qword ptr [rcx-0x08], 0
       jne      SHORT G_M27646_IG06
       mov      rcx, qword ptr [(reloc 0x40000000004203d0)]
       mov      rcx, gword ptr [rcx+0x08]
       call     Program:Consume(System.String)
       lea      r8, [(reloc 0x40000000004203c8)]
       add      r8, -8
       cmp      qword ptr [r8], 0
       jne      SHORT G_M27646_IG07
						;; size=47 bbWeight=1 PerfScore 14.25
G_M27646_IG03:  ;; offset=0x0034
       mov      rcx, qword ptr GS:[0x0058]
       mov      eax, dword ptr [(reloc 0x40000000004203d8)]
       mov      edx, (reloc 0x40000000004203e0)
       add      rdx, qword ptr [rcx+8*rax]
       mov      rbx, qword ptr [rdx]
       test     rbx, rbx
       je       SHORT G_M27646_IG08
						;; size=32 bbWeight=1 PerfScore 9.50
G_M27646_IG04:  ;; offset=0x0054
       mov      rcx, gword ptr [rbx+0x80]
       call     Program:Consume(System.String)
       nop      
						;; size=13 bbWeight=1 PerfScore 3.25
G_M27646_IG05:  ;; offset=0x0061
       add      rsp, 32
       pop      rbx
       ret      
						;; size=6 bbWeight=1 PerfScore 1.75
G_M27646_IG06:  ;; offset=0x0067
       call     CORINFO_HELP_READYTORUN_GCSTATIC_BASE
       mov      rcx, qword ptr [(reloc 0x40000000004203d0)]
       mov      rcx, gword ptr [rcx+0x08]
       call     Program:Consume(System.String)
       lea      r8, [(reloc 0x40000000004203c8)]
       add      r8, -8
       cmp      qword ptr [r8], 0
       je       SHORT G_M27646_IG03
						;; size=38 bbWeight=0 PerfScore 0.00
G_M27646_IG07:  ;; offset=0x008D
       xor      ecx, ecx
       mov      edx, -1
       lea      rax, [(reloc 0x40000000004203f0)]
       call     rax
       mov      rbx, rax
       jmp      SHORT G_M27646_IG04
						;; size=21 bbWeight=0 PerfScore 0.00
G_M27646_IG08:  ;; offset=0x00A2
       mov      rcx, rdx
       lea      rax, [(reloc 0x40000000004203e8)]
       call     rax
       mov      rbx, rax
       jmp      SHORT G_M27646_IG04
						;; size=17 bbWeig
```

I am not sure how common it is to see such patterns, @VSadov - any thoughts?

In cases such as below, we generate expected code:

```c#

    [ThreadStatic]
    static string i = 5.ToString();
    [ThreadStatic]
    static string j = 5.ToString();

        for (int k = 0; k < 10; k++)
        {            
            Consume(i);
            Consume(j);
        }
```


```asm
G_M27646_IG02:  ;; offset=0x0006
       xor      ebx, ebx
       lea      r8, [(reloc 0x40000000004203e8)]
       add      r8, -8
       cmp      qword ptr [r8], 0
       jne      SHORT G_M27646_IG05
       mov      rcx, qword ptr GS:[0x0058]
       mov      eax, dword ptr [(reloc 0x40000000004203c8)]
       mov      edx, (reloc 0x40000000004203d0)
       add      rdx, qword ptr [rcx+8*rax]
       mov      rsi, qword ptr [rdx]
       test     rsi, rsi
       je       SHORT G_M27646_IG06
                                                ;; size=51 bbWeight=1 PerfScore 14.50
G_M27646_IG03:  ;; offset=0x0039
       mov      rcx, gword ptr [rsi+0x68]
       call     Program:Consume(System.String)
       mov      rcx, gword ptr [rsi+0x78]
       call     Program:Consume(System.String)
       inc      ebx
       cmp      ebx, 10
       jl       SHORT G_M27646_IG03
```

Reply:
Of general patterns, I think we can only assume that threadstatics are relatively less common than other kinds of fields. Also it is not common to have a lot of them in the same type. It is valid and certainly possible to see numerous threadstatics in a type, but it is more common for a type to have just one such field that holds all the threadstatic state for the type. 
I'd say - 2 threadstatic fields in a type is likely common enough, 20 - is much less likely.

Mixing threadstatic and regular static fields in the same type is fairly common. 

As for the order of access within the same method - static or threadstatic first, I think it would be too subtle if one pattern performs better than another.

Reply:
> I think we can only assume that threadstatics are relatively less common than other kinds of fields

On the other hand, I'd guess that code that uses them is often performance sensitive so handling it efficiently for AOT is important.